File: tools/compile-pinned.mjs
// Dump every positive's PINNED GLSL, so a real compiler can see it.
//
// Everything checked so far is a scan over text. Text scans do not
// catch a missing function, a type that does not line up, or a name
// that collides with a GLSL keyword — and pinning replaced a dozen
// builtins with functions of the engine's own, any of which could be
// spelled wrong in a way no regex notices.
//
// So this writes what a driver would actually be handed:
//
//   detlib.glsl   the proven library, byte-identical to the darkroom's
//   detpre.glsl   the engine's pinned helpers, built on it
//   <id>.glsl     the emitted shape function
//
// tools/compile-pinned.py then compiles each one on a real context.
// Splitting it in two is not tidiness — node has no GL, and the
// darkroom already owns a working moderngl harness, so the join is a
// directory of files rather than a binding.
//
//   node tools/compile-pinned.mjs   -> build/pinned/

import { readdirSync, writeFileSync, mkdirSync, readFileSync } from "node:fs";
import { fileURLToPath, pathToFileURL } from "node:url";
import { dirname, join, resolve } from "node:path";
import { emitWalk } from "../core/emit.mjs";
import { substitute } from "../core/oracle.mjs";
import { unfuse } from "../core/unfuse.mjs";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..");
const OUT = join(ROOT, "build", "pinned");
mkdirSync(OUT, { recursive: true });

// The det library is filled from the record and then UNFUSED, which is
// how the darkroom ships it (core/unfuse.mjs): every fma() in the
// generator's source becomes a multiply and an add, because that is
// the form every measured stack computes identically. The prelude is
// this engine's own and keeps its fma calls, which the emitted plates
// also carry; the darkroom bakes both the way it bakes plates.
for (const [src, dst, rewrite] of [["detlib.glsl.template", "detlib.glsl", true],
                                   ["detpre.glsl.template", "detpre.glsl", false]]) {
  const t = readFileSync(join(ROOT, "core", src), "utf8");
  const filled = substitute(t);
  writeFileSync(join(OUT, dst), rewrite ? unfuse(filled).text : filled, "utf8");
}

// BOTH VARIANTS, because a bit-identity result without a control is
// not a result. If pinned plates agree across two drivers, the obvious
// question is whether ANY plate would - maybe the sample set is too
// small, maybe these plates are arithmetically dull. Emitting the same
// positives unpinned and running them the same way answers it: the
// unpinned run must DISAGREE where the pinned run agrees, or the
// experiment has no power and the agreement means nothing.
const UNP = join(ROOT, "build", "unpinned");
mkdirSync(UNP, { recursive: true });

const POS = join(ROOT, "positives");
let ok = 0, refused = 0, ctrl = 0;
const skipped = [];
for (const f of readdirSync(POS).filter(x => x.endsWith(".pos.mjs")).sort()) {
  const pos = (await import(pathToFileURL(resolve(join(POS, f))).href)).default;
  const id = f.replace(".pos.mjs", "");
  try {
    writeFileSync(join(OUT, `${id}.glsl`), emitWalk(pos, { pin: true }),
                  "utf8");
    ok++;
    // the control only exists for plates the pinned set accepted, so
    // the two sets are the same plates and the comparison is paired
    writeFileSync(join(UNP, `${id}.glsl`), emitWalk(pos), "utf8");
    ctrl++;
  } catch (e) {
    refused++;
    skipped.push(id);
  }
}
console.log(`build/pinned:   detlib.glsl, detpre.glsl, and ${ok} plates`);
console.log(`build/unpinned: ${ctrl} of the same plates, as the control`);
console.log("  now: python tools/compile-pinned.py --run <tag>");

// A REFUSAL IS A REGRESSION HERE, whatever it is elsewhere. Refusing an
// unpinnable construct is the design working when somebody is writing a
// positive; in a corpus where all sixty-nine emit, a plate that stops
// emitting is a plate that stops existing, and this printed it as a
// note and exited 0. Nothing downstream would have noticed: bakeemitted
// reads build/pinned/<id>.glsl, and a file that was not rewritten this
// run is a file left over from the last one - which is precisely the
// staleness its own guard exists to refuse.
if (refused) {
  console.log(`\n  ${refused} REFUSED by the pinned set: ${skipped.join(" ")}`);
  console.log("  build/pinned still holds whatever the last successful run");
  console.log("  left for those, so nothing here is safe to bake.");
  process.exit(1);
}

File: tools/compile-pinned.py
#!/usr/bin/env python3
"""Compiles every pinned plate on a real driver.

Everything else in Phase 2 is a scan over text, and a scan cannot see a
missing function, a mismatched type, or a name that collides with a
GLSL keyword. Pinning swapped a dozen builtins for functions the engine
wrote itself, so exactly those mistakes are the ones it could make.

The prelude is taken from the DARKROOM, not written here: pal, hashu,
u2f and the complex helpers are the registry's, and a plate emitted
against a private copy of them would compile here and fail in
PrettyCloud. Extracted by the same regexes shapeprobe.py uses, so if
the camera moves this stops rather than drifting.

    python tools/compile-pinned.py [--stack N]
"""
import argparse
import os
import pathlib
import re
import sys

HERE = pathlib.Path(__file__).resolve().parent
ROOT = HERE.parent
PINNED = ROOT / "build" / "pinned"
DARKROOM = pathlib.Path(os.environ.get("DARKROOM", ROOT.parent /
                                       "atlas-darkroom"))


def prelude():
    src = (DARKROOM / "darkroom" / "shader.py").read_text(encoding="utf-8")
    want = [r"const float PI  = [0-9.]+;",
            r"const float TAU = [0-9.]+;",
            r"vec3 pal\(float t, vec3 a, vec3 b, vec3 c, vec3 d\)\{.*?\n\}",
            r"uint hashu\(uint x\)\{.*?\n\}",
            r"float u2f\(uint x\)\{[^\n]*\}",
            r"vec2 cmul\(vec2 a, vec2 b\)\{[^\n]*\}",
            r"vec2 cdiv\(vec2 a, vec2 b\)\{[^\n]*\}",
            r"vec2 cinv\(vec2 a\)\{[^\n]*\}"]
    out = []
    for pat in want:
        m = re.search(pat, src, re.S)
        if not m:
            sys.exit(f"prelude piece not found in shader.py: {pat}\n"
                     f"the camera moved; this would have compiled a "
                     f"program nothing renders")
        out.append(m.group(0))
    # csqrt is a multi-line helper in the header; take it whole
    m = re.search(r"vec2 csqrt\(vec2 z\)\{.*?\n\}", src, re.S)
    if m:
        out.append(m.group(0))
    return "\n".join(out)


# uT IS PART OF THE PLATE CONTRACT, not of this harness. Animated
# plates read it, the darkroom sets it per exposure from uT0 and the
# shutter, and a compile without it fails on eighteen plates for a
# reason that has nothing to do with pinning. Declared here so the
# check measures what it claims to.
HEAD = """#version 430
layout(local_size_x = 1) in;
layout(std430, binding = 0) writeonly buffer Out { float o[]; };
uniform float uT;
"""

TAIL = """
void main(){
  float P[8];
  for (int k = 0; k < 8; k++) P[k] = float(k) * 0.125;
  vec3 col;
  vec3 v = %s(vec2(0.25, 0.5), vec4(0.1, 0.2, 0.3, 0.4), 12345u, P, col);
  o[0] = v.x + col.x;
}
"""


def main(argv=None):
    ap = argparse.ArgumentParser(prog="compile-pinned")
    ap.add_argument("--stack", type=int, default=None,
                    help="EGL device index; omit for the default context")
    a = ap.parse_args(argv)

    if not PINNED.exists():
        sys.exit("no build/pinned - run `node tools/compile-pinned.mjs`")
    import moderngl
    kw = {"standalone": True, "require": 430}
    if a.stack is not None:
        kw.update(backend="egl", device_index=a.stack)
    ctx = moderngl.create_context(**kw)
    print(f"  {ctx.info['GL_RENDERER'][:60]}")

    pre = prelude()
    detlib = (PINNED / "detlib.glsl").read_text(encoding="utf-8")
    detpre = (PINNED / "detpre.glsl").read_text(encoding="utf-8")
    base = HEAD + pre + "\n" + detlib + "\n" + detpre + "\n"

    # last run's failures are not this run's input
    for stale in PINNED.glob("*.FAILED.glsl"):
        stale.unlink()
    plates = sorted(p for p in PINNED.glob("*.glsl")
                    if p.name not in ("detlib.glsl", "detpre.glsl")
                    and not p.name.endswith(".FAILED.glsl"))
    good, bad = [], []
    for p in plates:
        body = p.read_text(encoding="utf-8")
        m = re.search(r"vec3 (shape_[A-Za-z0-9_]+)\(", body)
        if not m:
            bad.append((p.stem, "no shape_ entry point"))
            continue
        src = base + body + TAIL % m.group(1)
        try:
            ctx.compute_shader(src)
            good.append(p.stem)
        except Exception as exc:                   # noqa: BLE001
            # the useful part of a driver's log is the lines with
            # "error" in them, not the three-line banner in front
            lines = [ln.strip() for ln in str(exc).splitlines()
                     if "error" in ln.lower() and ln.strip()]
            bad.append((p.stem, " | ".join(lines[:2])[:200]
                        or str(exc).strip()[:200]))
            (PINNED / f"{p.stem}.FAILED.glsl").write_text(src,
                                                          encoding="utf-8")

    print(f"\n  compiled: {len(good)}/{len(plates)}")
    if bad:
        print(f"  FAILED  : {len(bad)}")
        for name, why in bad:
            print(f"    {name:14} {why}")
        print(f"\n  full sources written beside them as *.FAILED.glsl")
    else:
        print("  every pinned plate compiles, det_ helpers and all")
    try:
        ctx.release()
    except Exception:                              # noqa: BLE001
        pass
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())
